fix(linter): improve no-zero-fractions rule for member expressions and scientific notation

cblh opened the pull request:
fix: #4788

- Added a check for parent nodes that are MemberExpressions
- If a parent node is a MemberExpression, the fixed number is now wrapped in parentheses to maintain correct syntax
- Add support for scientific notation (e.g., 1e10, 1.e10)
- Handle special cases where a space is needed after certain keywords (return, throw, typeof, void) to prevent the number from being interpreted as a property access

cblh committed:
fix(linter): improve no-zero-fractions rule for member expressions

- Added a check for parent nodes that are MemberExpressions
- If a parent node is a MemberExpression, the fixed number is now wrapped in
  parentheses to maintain correct syntax

cblh committed:
feat(linter): improve no-zero-fractions rule

This commit enhances the no-zero-fractions rule in the unicorn linter:

- Add support for scientific notation (e.g., 1e10, 1.e10)
- Improve handling of member expressions
- Introduce is_decimal_integer utility function
- Update test cases and snapshots to reflect new behavior
- Fix various edge cases in number formatting

These changes make the rule more robust and consistent with the original
ESLint implementation, especially for scientific notation and member
expressions.

cblh committed:
chore(linter): Refactor `NoZeroFractions` rule for improved formatting and clarity

- Simplified the condition that checks for decimal integers within member expressions by removing unnecessary parentheses around the formatted string.
- Updated string formatting in `format_raw` function to use the shorthand `{variable}` syntax instead of `{}` with positional arguments.

cblh committed:
fix(linter): improve no-zero-fractions rule

- Handle special cases where a space is needed after certain keywords
  (return, throw, typeof) to prevent the number from being interpreted
  as a property access
- Add tests for these special cases
- Update existing tests to cover more scenarios

cblh committed:
fix(linter): add 'void' keyword support to no-zero-fractions rule

- Include 'void' in the list of keywords that require a space before
  the number to prevent misinterpretation as property access
- Add test cases for 'void' keyword scenarios

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_zero_fractions.rs:
```diff
@@ -146,22 +177,28 @@ fn test() {
         (r"const foo = (1.).toString()", r"const foo = (1).toString()"),
-        // (r"1.00.toFixed(2)", r"(1).toFixed(2)"),
-        // (r"1.00 .toFixed(2)", r"(1) .toFixed(2)"),
+        (r"1.00.toFixed(2)", r"(1).toFixed(2)"),
```
Could this break ASI?
```js
console.log('foo')
1.00.toFixed(2) // -> (1).toFixed(2), parsed as a function call
```

cblh replied:
i will fix it

cblh replied:
I need help, I don't know how to get the previous token.

DonIsaac replied:
this case is also a problem blocking #4312. I'll add a task to add a `delete_statement` method to `RuleFixer`.

cblh replied:
Sorry, I don't know what `delete_statement` does? The logic I need to implement here is It : 

checks the type and value of tokenBefore:
If tokenBefore is a Punctuator (e.g., a symbol like ;, ], or )), it determines whether a semicolon is necessary based on the context (e.g., the type of the last block node).
If the token type is in tokenTypesNeedsSemicolon, it returns true (semicolon needed).
Special cases like Template strings, ObjectExpression blocks, and certain Identifier cases are handled explicitly. 

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/5c92a9aed7793afa00f23e72e25bc58c10adecd4/rules/utils/needs-semicolon.js#L33

Is this also a blocking problem or is there an API available?

DonIsaac replied:
This API does not exist yet, we should move ahead with this improvement for the time being.

cblh replied:
> This API does not exist yet, we should move ahead with this improvement for the time being.

Since it is not a problem with the current rules, let’s merge this PR first. Then join a new issue to follow up on this API

DonIsaac commented on crates/oxc_linter/src/rules/unicorn/no_zero_fractions.rs:
```diff
@@ -66,12 +66,43 @@ impl Rule for NoZeroFractions {
             },
-            |fixer| fixer.replace(number_literal.span, fmt),
+            |fixer| {
+                let mut fixed = fmt.clone();
```
Why `clone()` instead of a move?

cblh replied:
you are right

DonIsaac commented on crates/oxc_linter/src/utils/unicorn/numeric.rs:
```diff
@@ -0,0 +1,13 @@
+// Copied from https://github.com/eslint/eslint/blob/cc4871369645c3409dc56ded7a555af8a9f63d51/lib/rules/utils/ast-utils.js#L1237
+lazy_static! {
+    static ref DECIMAL_INTEGER_PATTERN: Regex =
+        Regex::new(r"^(?:0|0[0-7]*[89]\d*|[1-9](?:_?\d)*)$").unwrap();
```
Does `NumericLiteral#base` not suit your needs?

cblh replied:
This is to check whether 1.0100 becomes 1.01 or 1.00 becomes 1. I found that format_raw's return is_dangling_dot can distinguish this situation, so I will delete this code

cblh committed:
refactor(linter): remove `is_decimal_integer` utility and simplify `no-zero-fractions` rule

- Removed the `is_decimal_integer` function and related `numeric.rs` utility, as it was no longer necessary.
- Simplified the `no-zero-fractions` rule by directly checking for non-dangling dots instead of using the removed utility.
- Updated the logic for fixing number formatting to remove redundant checks.

cblh committed:
chore(linter): fmt

cblh committed:
fix(unicorn/no-zero-fractions): improve handling of member expressions and decimal integers

- Refactored logic in the `NoZeroFractions` rule to ensure proper handling of cases where the numeric value is a decimal integer and part of a member expression.
- Added a clone of the format string to avoid mutating the original `fmt`.
- Introduced checks for `is_decimal_integer` and `is_member_expression` to better control the formatting of the fixed value.
- Updated test cases to cover additional scenarios, including fixing fractional values like `1.010.toFixed(2)` to `1.01.toFixed(2)`.

DonIsaac committed:
Merge branch 'main' into fix/4788

DonIsaac committed:
Merge branch 'main' into fix/4788

cblh commented:
> This API does not exist yet, we should move ahead with this improvement for the time being.

Since it is not a problem with the current rules, let’s merge this PR first. Then join a new issue to follow up on this API

DonIsaac committed:
Merge branch 'main' into fix/4788

DonIsaac commented:
Is there still a service disruption for Github Actions?

cblh committed:
Merge branch 'main' into fix/4788

cblh commented:
> Is there still a service disruption for Github Actions?

@DonIsaac github action service has been restored

cblh committed:
feat(linter): Enhance no_zero_fractions rule handling

- Add TODO for more complex token checks and semicolon handling
- Prepare for special keyword cases requiring space after fixed value

cblh committed:
feat(linter): Enhance no_zero_fractions rule handling

- Add TODO for more complex token checks and
